relay-server/src/utils/squeue.rs

```diff
@@ -0,0 +1,190 @@
+
+impl<T> Stream for ScheduledQueue<T> {
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.waker = Some(cx.waker().clone());
+
+        let sleep = self.sleep.as_mut().poll(cx);
+
+        // No item means, we're pending until scheduled (through adding an item).
+        if let Some(next) = self.inner.peek() {
+            let when = next.when;
+
+            // TODO: maybe optimize here for `when` being in the past but the sleep
+            // is still in the future. This can happen when an item with a shorter
+            // deadline, than the item at the front of the queue, is inserted.
+            //
+            // With the current behaviour we will reset the sleep deadline (in the else branch)
+            // to a value which will immediately be ready.
+            // Only after the sleep yields ready, we yield the item.
+            //
+            // This is one more wakeup and sleep poll than necessary.
+            //
+            // The current design is much simpler and less prone to mistakes since
+            // the sleep is always synchronized to the first item in the queue.
+            if matches!(sleep, Poll::Ready(_)) && when <= Instant::now() {
+                // We already expired the first item, yield it.
+                let current = self.inner.pop().unwrap();
+
+                let next_deadline = self
+                    .inner
+                    .peek()
+                    .map(|item| item.when)
+                    .unwrap_or_else(far_future);
+                self.sleep.as_mut().reset(next_deadline);
+                // Immediately wake up to check the next item and to poll the new sleep deadline.
+                cx.waker().wake_by_ref();
+
+                return Poll::Ready(Some(current.value));
+            } else if self.sleep.deadline() != when {
+                // Somehow the deadline does not match the first item anymore, adjust the deadline.
+                // This may happen when there is a new item in the queue and we got woken to
+                // update the deadline or in the rare case where the first item got removed
+                // while we were getting woken from the sleep.
+                self.sleep.as_mut().reset(when);
+                // Immediately wake to await the new deadline.
+                cx.waker().wake_by_ref();
+            }
+        }
+
+        // Next wake up already triggered or will be triggered by sleep.
+        Poll::Pending
+    }
+}
+
+impl<T> FusedStream for ScheduledQueue<T> {
+    fn is_terminated(&self) -> bool {
+        // The stream never returns `Poll::Ready(None)`.
+        false
+    }
+}
+
+impl<T> Default for ScheduledQueue<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+struct Item<T> {
+    when: Instant,
+    value: T,
+}
+
+impl<T> PartialEq for Item<T> {
+    fn eq(&self, other: &Self) -> bool {
+        other.when == self.when
```

Comment:
nit: just to have it all in one place.

```suggestion
        self.cmp(other).is_eq()
```

Reply:
This goes through the `Ord` trait, but this is the impl for `PartialEq` which *is* used for `==`, the other two impls are for `{Partial,}Ord`.